### ISSUE TITLE
Generate the Message-ID earlier

### ISSUE THREAD

**Issue by alejandro-colomar** [neomutt/neomutt#4382]
Hi!

I'd like to be able to cryptographically sign the Message-ID, since it's more or less important information (we do sign the In-Reply-To header, so it would be consistent to also sign this).  thunderbird(1) does sign it.

However, for that to be possible, neomutt(1) should generate the Message-ID in an earlier stage, before the crypto code runs.

It would be interesting if it were generated before opening the editor, so that I already see in the editor what the Message-ID will be.  git-format-patch(1) for example generates a Message-ID (if you use `--thread`), and sending such patches with neomutt(1) works fine, so technically it should just work.

Related: <https://github.com/neomutt/neomutt/discussions/4251>

Have a lovely day!
Alex

**Comment by alejandro-colomar** [comment]
@gahr 

I think I need to insert it here:

```diff
diff --git i/send/send.c w/send/send.c
index f697f966e..b60c0e041 100644
--- i/send/send.c
+++ w/send/send.c
@@ -350,6 +350,9 @@ static int edit_envelope(struct Envelope *en, SendFlags flags, struct ConfigSubs
     goto done;
   }
   mutt_env_set_subject(en, buf_string(buf));
+
+  // Set Message-ID
+
   rc = 0;
 
 done:
```

Would you mind telling me where it is currently being generated?

**Comment by gahr** [comment]
It's here: 
https://github.com/neomutt/neomutt/blob/74c7004ec3b89daccdebd3d56d2fd28f800383e3/send/sendlib.c#L804

**Comment by ossilator** [comment]
another reason for generating the msgid early would be postponed messages gaining one. this would aid some forms of offline-imap synchronization.

**Comment by gahr** [comment]
And there's another call to `gen_msgid` in the `bounce` path.

**Comment by gahr** [comment]
I am still not totally convinced that it's a good idea to generate the msg-id way before send time, as it would be in the case of a postponed msg.

Anyway, if you have `edit_headers`, you *can* already set your own `message-id` header.

**Comment by alejandro-colomar** [comment]
I'll prepare the patch, and we can debate what are the pros and cons about it in parallel.  What do you think can go wrong by generating it early?

**Comment by gahr** [comment]
probably nothing, as we're now generating it randomly. But IIRC there's been people asking for a more descriptive message id including timestamps or server names. Timestamps referring to a time other than the send date would be a bit weird... not so big of a deal, probably.

**Comment by alejandro-colomar** [comment]
Draft:

```dif
diff --git i/send/send.c w/send/send.c
index f697f966e..f7c890b57 100644
--- i/send/send.c
+++ w/send/send.c
@@ -350,6 +350,10 @@ static int edit_envelope(struct Envelope *en, SendFlags flags, struct ConfigSubs
     goto done;
   }
   mutt_env_set_subject(en, buf_string(buf));
+
+  if (!en->message_id)
+    env->message_id = gen_msgid();
+
   rc = 0;
 
 done:
diff --git i/send/sendlib.c w/send/sendlib.c
index b0ba0070d..f7eff3896 100644
--- i/send/sendlib.c
+++ w/send/sendlib.c
@@ -729,7 +729,7 @@ const char *mutt_fqdn(bool may_hide_host, const struct ConfigSubset *sub)
 }
 
 /**
- * gen_msgid - Generate a random Message ID
+ * mutt_gen_msgid - Generate a random Message ID
  * @retval ptr Message ID
  *
  * The length of the message id is chosen such that it is maximal and fits in
@@ -751,7 +751,7 @@ const char *mutt_fqdn(bool may_hide_host, const struct ConfigSubset *sub)
  *
  * @note The caller should free the string
  */
-static char *gen_msgid(void)
+char *mutt_gen_msgid(void)
 {
   const int ID_LEFT_LEN = 50;
   const int ID_RIGHT_LEN = 12;
@@ -802,7 +802,7 @@ void mutt_prepare_envelope(struct Envelope *env, bool final, struct ConfigSubset
     mutt_set_followup_to(env, sub);
 
     if (!env->message_id)
-      env->message_id = gen_msgid();
+      env->message_id = mutt_gen_msgid();
   }
 
   /* Take care of 8-bit => 7-bit conversion. */
@@ -875,7 +875,7 @@ static int bounce_message(FILE *fp, struct Mailbox *m, struct Email *e,
     fprintf(fp_tmp, "Resent-Date: %s\n", buf_string(date));
     buf_pool_release(&date);
 
-    char *msgid_str = gen_msgid();
+    char *msgid_str = mutt_gen_msgid();
     fprintf(fp_tmp, "Resent-Message-ID: %s\n", msgid_str);
     FREE(&msgid_str);
     mutt_addrlist_write_file(to, fp_tmp, "Resent-To");
diff --git i/send/sendlib.h w/send/sendlib.h
index 3254814fd..5a41e8b2d 100644
--- i/send/sendlib.h
+++ w/send/sendlib.h
@@ -36,6 +36,7 @@ struct Mailbox;
 
 int              mutt_bounce_message     (FILE *fp, struct Mailbox *m, struct Email *e, struct AddressList *to, struct ConfigSubset *sub);
 const char *     mutt_fqdn               (bool may_hide_host, const struct ConfigSubset *sub);
+char *           mutt_gen_msgid          (void);
 enum ContentType mutt_lookup_mime_type   (struct Body *b, const char *path);
 struct Body *    mutt_make_file_attach   (const char *path, struct ConfigSubset *sub);
 struct Body *    mutt_make_message_attach(struct Mailbox *m, struct Email *e, bool attach_msg, struct ConfigSubset *sub);
```

**Comment by gahr** [comment]
Would you please open a draft PR? It'd be so much easier to review / test than a diff in the comments :) thanks!

**Comment by alejandro-colomar** [comment]
> probably nothing, as we're now generating it randomly. But IIRC there's been people asking for a more descriptive message id including timestamps or server names. Timestamps referring to a time other than the send date would be a bit weird... not so big of a deal, probably.

The timestamp, well, it will signal that the message was generated a bit earlier than the actual sending.  It already happens with the Date field and the PGP signature date, which necessarily happen at slightly different times.
It would still be relatively coherent (with postponed messages, that difference becomes bigger, but it shouldn't be a huge problem).  That should only be for making it unique.

As for the server name, I guess you refer to @nabijaczleweli .  I say some issue opened by it when checking if this had been discussed.  I don't know what to say.  :)

**Comment by alejandro-colomar** [comment]
> Would you please open a draft PR? It'd be so much easier to review / test than a diff in the comments :) thanks!

Sure, that was my plan.  This is just the trailer.  :)
